### PR TITLE
[V-1] VRF foundation

### PR DESCRIPTION
**Pull request by KyryloKupin**

### CODE REVIEW

File: src/vrf.c [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    free(verdict);
+    verdict = NULL;
+}
+
+int
+verify(Vrf **result)
+{
+    if (extract_local_part_and_domain(result) == EXIT_FAILURE) {
+        printf("Email parts extraction failure.");
+        return EXIT_FAILURE;
```
Comment (irog-c) [inline src/vrf.c]: I believe that EXIT_SUCCESS and EXIT_FAILURE are used as a program execution status, not sure if it is used in the right context here in this function? [Source](https://en.cppreference.com/w/c/program/EXIT_status)

Comment (KyryloKupin) [reply src/vrf.c]: Agree

File: include/vrf.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+//
+// Created by Tuposoft Collective on 23.01.2023.
+//
+
+#pragma once
+
+#define SMTP_DATA_LINES_MAX_LENGTH 998
+
+#include <stdbool.h>
```
Comment (irog-c) [inline include/vrf.h]: Do we need this included here, I don't see any boolean usage in this file? We should probably move it to the .c file.

File: src/vrf.c [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    char status[4];
+    memcpy(status, buffer, 3);
+    status[3] = '\0';
+    long code = strtol(status, NULL, 0);
+    (*result)->result = code == 250;
+
+    send_command(sock, "QUIT\n");
+    close(client_fd);
+
+    return EXIT_SUCCESS;
```
Comment (irog-c) [inline src/vrf.c]: Nitpick: probably should remove this as well, and return `VRF_OK`.

File: src/vrf.c [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    VRF result = malloc(sizeof *result);
+    if (errno) return EXIT_FAILURE;
+    if (!(result->email = strdup(argv[1]))) return EXIT_FAILURE;
+    err = verify(&result);
+    if (err == VRF_ERR) return EXIT_FAILURE;
+    err = print_vrf(stdout, result);
+    if (err == VRF_ERR) return EXIT_FAILURE;
+    free_vrf(result);
+
+    return 0;
```
Comment (irog-c) [inline src/vrf.c]: Return `EXIT_SUCCESS`?

File: src/vrf.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+static VRF_err
+read_response(int sock, char *buffer)
+{
+    char (*b)[SMTP_DATA_LINES_MAX_LENGTH] = (char (*)[SMTP_DATA_LINES_MAX_LENGTH]) buffer;
+    if (read(sock, *b, sizeof *b) < 0) {
+        printf("Failed to read from socket.\n");
+        return VRF_ERR;
+    }
+    if (PRINT_RESPONSE) {
```
Comment (irog-c) [inline src/vrf.c]: Is `PRINT_RESPONSE` a macro? If yes we can use `#ifdef`, so that it is done at compile-time.

**Comment by irog-c** [comment]
I think we should add a `.gitignore` file.

File: src/vrf.c [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    bool catch_all;
+};
+
+static VRF_err_t
+send_command(int sock, char *format, ...)
+{
+    va_list args;
+    va_start(args, format);
+    char *command;
+    if (vasprintf(&command, format, args) < 0) {
```
Comment (irog-c) [inline src/vrf.c:46]: Should we really use `vasprintf`? Judging by the documentation [here](https://linux.die.net/man/3/vasprintf), it says that it allocates memory, but you never free it so I think this causes a memory leak.

Comment (irog-c) [reply src/vrf.c:46]: Oh never mind, you free it, sorry :smiley:

**Comment by irog-c** [comment]
Suggestion:
Instead of doing `DEBUG_RESPONSE`, we can make a define that looks something like this:
```
#ifdef DEBUG
    #define DEBUG_PRINT(...) printf(__VA_ARGS__)
#else
    #define DEBUG_PRINT(x) x
#endif
```
What do you think @KyryloKupin ?
That way we don't have to do an #ifdef every time we want to check if it's defined, because if it isn't it's just gonna put nothing and it'll be an "empty" macro resolution.

**Comment by KyryloKupin** [comment]
> Suggestion: Instead of doing `DEBUG_RESPONSE`, we can make a define that looks something like this:
> 
> ```
> #ifdef DEBUG
>     #define DEBUG_PRINT(...) printf(__VA_ARGS__)
> #else
>     #define DEBUG_PRINT(x) x
> #endif
> ```
> 
> What do you think @KyryloKupin ? That way we don't have to do an #ifdef every time we want to check if it's defined, because if it isn't it's just gonna put nothing and it'll be an "empty" macro resolution.

Feel free to branch off, code your suggestions and create a PR.

**Comment by irog-c** [comment]
> > Suggestion: Instead of doing `DEBUG_RESPONSE`, we can make a define that looks something like this:
> > ```
> > #ifdef DEBUG
> >     #define DEBUG_PRINT(...) printf(__VA_ARGS__)
> > #else
> >     #define DEBUG_PRINT(x) x
> > #endif
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What do you think @KyryloKupin ? That way we don't have to do an #ifdef every time we want to check if it's defined, because if it isn't it's just gonna put nothing and it'll be an "empty" macro resolution.
> 
> Feel free to branch off, code your suggestions and create a PR.

Done, here is my [PR](https://github.com/TupoSoft/vrf/pull/2).

**Review by irog-c — APPROVED** [review]

**Pull request merged**